### ISSUE TITLE
Parser - Implement Collection Segment/ Grammar

### ISSUE THREAD

**Issue by WittierDinosaur** [sqlfluff/sqlfluff#3386]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Currently, our StatementSegment()'s matching grammar is a OneOf(), wrapping every statement type that we support. These matchers then scan each listed segment in order, in an effort to find a match. This is suboptimal. If, for example, the Statement is 
```GRANT USAGE ON SCHEMA foo TO bar;```
there is no point scanning N many different statements that start with `CREATE`. To solve this, we could have a Collection(). These are statements (though there's no intrinsic reason it couldn't apply to lower grammars) that act as like a OneOf(), except that they are only assessed if and only if their starting token matches. Example usage:

```
CreateStatement = Collection(
    CreateFunctionStatementSegment,
    CreateTableStatementSegment,
    CreateUserStatementSegment,
    ...
    token = 'CREATE'
)

class StatementSegment:

match_grammar OneOf(
    CreateStatement,
    ...
)
```

### Dialect

N/A

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
While this seems sensible, are we sure it’s an issue? Or is SQLFluff perhaps efficient enough to stop after the mismatching the first keyword (that this optimisation implies) so checking 5 or 10 `CREATE` statements is no slower than adding a level of indirection? Don’t know the answer btw, just asking the question.

**Comment by WittierDinosaur** [comment]
I don't think this is the biggest thing in terms of performance, but it's definitely net beneficial. Consider a dialect with 200 different statements but only 15 entrypoints, and a file which is a set of statements which are 190th in the list. That's 189 attempts to match the statement before it hits gold, which is significantly more than 15 + subset
